### PR TITLE
New type field

### PR DESCRIPTION
**Pull request by josojo**
This PR would pass all tests, if pepper recognizes new field type. PR submitted

With the new type: field, which has exactly the same properties as the prime field of the alt-bn-curve, we can simulate the snarks logic much closer to the real logic in cpp. 
This allows better testing.

### CODE REVIEW

File: pepper/test/apps/isBoolVerification_test.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#include <gtest/gtest.h>
+#include "declarations.h"
+#include <apps/hash_transform.h>
+
+#include <apps/dex_common.h>
+#include "util.h"
+
+struct TestStructArray { field digest[256]; };
+
+TEST(isBoolVerification, isBoolVerificationWithBools) {
```
Comment (fleupold) [inline pepper/test/apps/isBoolVerification_test.cpp]: Ideally this wouldn't be part of this pull request (it's kind of unrelated).

File: pepper/Makefile [hunk 2]
```diff
@@ -43,15 +43,20 @@ link-exo: check-env
 	ln -sf ~/dex-zksnarks/exo_compute/private_input.py $(PEPPER_BIN)/exo0
 
 # Target to run all tests (e.g. used by travis)
-test: unit-tests gadget-tests e2e-tests
+test: unit-tests-iBV unit-tests gadget-tests e2e-tests
```
Comment (fleupold) [inline pepper/Makefile]: Let's keep this `test` formula small and rather have `unit-tests` be a target that lists all possible unit tests sub-targets (maybe there is a way we can do this in one formula with a loop for each `_test.cpp` file in test/apps/

File: pepper/Makefile [hunk 3]
```diff
@@ -43,15 +43,20 @@ link-exo: check-env
...
 
 e2e-tests: check-env
 	test/e2e/hash_transform.sh
 
 unit-tests: test/apps/hash_transform_test.cpp
-	$(CXX) -DTEST $(TEST_IFLAGS) $< -o $(TEST_BIN)/run_unit_tests $(TEST_LDFLAGS)
+	$(CXX) $(CXXFLAGS) -DTEST $(TEST_IFLAGS) $(IFLAGS) $< -o $(TEST_BIN)/run_unit_tests $(TEST_LDFLAGS) $(LDFLAGS) 
 	$(TEST_BIN)/run_unit_tests
 
+unit-tests-iBV: test/apps/isBoolVerification_test.cpp
```
Comment (fleupold) [inline pepper/Makefile]: let's please not use abbreviations like this. It might save you 15 characters typing now, but it will make it harder to see what this is testing without also finding the test file name (which if we start like this might become abbreviated as well).

**Review by fleupold — COMMENTED** [review]
Nothing big, just some feedback on structuring the makefile. Also, I'd really appreciate if we could have different pull requests for different features (e.g. one for supporting field types, and a separate one for testing the isBoolVerification).

File: pepper/Makefile (hunk 3 above)
Comment (josojo) [reply pepper/Makefile]: you are right

File: pepper/test/apps/isBoolVerification_test.cpp (hunk 1 above)
Comment (josojo) [reply pepper/test/apps/isBoolVerification_test.cpp]: removed it

File: pepper/Makefile (hunk 2 above)
Comment (josojo) [reply pepper/Makefile]: recursive testing all files is done

**Review by fleupold — COMMENTED** [review]
Can you revert the changes to the makefile? Those ones should be in the PR that creates the new test case.

Otherwise, looks great! Now we have to wait for pepper to be updated?

**Comment by josojo** [comment]
Finally, I also undid the changes in the makefile for the testing. (actually, I wanted to delete the one commit introducing this change. But since I did not find a good way to do that quickly, I simply pushed a new commit undoing the changes. Hopefully, this will not remove the changes from the other pr. I am not sure about it.)

Maybe we can merge this one rather earlier than sooner. Does not make sense that we keep on working with uint128. 
If @fleupold you could update your docker with the the pepper supporting field, then also we could test all of this pr.

**Comment by fleupold** [comment]
I'd prefer not diverging from pepper master unless we really need to. If there were comments/changes on your pull request (such as a change in the type name), we'd have to go through the procedure twice.

Let's wait if the PR is merged by Monday and otherwise base the docker image off our branch? I can also ping Max to see if he has any concerns merging this PR. Would that be ok for you?

**Comment by josojo** [comment]
sounds good

File: pepper/Makefile [hunk 4]
```diff
@@ -14,6 +14,7 @@ GADGET_SOURCES = ext_gadget/gadgets/common.cpp ext_gadget/gadgets/jubjub/*.cpp
 TEST_LDFLAGS = -lgtest -lpthread
 TEST_IFLAGS = -I./
 TEST_BIN = test/bin/
+FILENAMETESTS := $(shell find test/apps/ -name '*_test.cpp')
```
Comment (fleupold) [inline pepper/Makefile]: These changes seem unrelated.

File: pepper/Makefile [hunk 5]
```diff
@@ -51,7 +52,7 @@ e2e-tests: check-env
 unit-tests: $(patsubst test/apps/%.cpp, app_%, $(wildcard test/apps/*_test.cpp))
 
 app_%_test: test/apps/%_test.cpp
-	$(CXX) $(CXXFLAGS) -DTEST $(TEST_IFLAGS) $< -o $(TEST_BIN)/$@ $(TEST_LDFLAGS)
+	$(CXX) $(CXXFLAGS) -DTEST $(TEST_IFLAGS) $(IFLAGS) $< -o $(TEST_BIN)/$@ $(TEST_LDFLAGS) $(LDFLAGS)
```
Comment (fleupold) [inline pepper/Makefile]: ditto

File: .travis.yml [hunk 6]
```diff
@@ -2,10 +2,15 @@ sudo: required
 services:
 - docker
 install:
+<<<<<<< HEAD
```
Comment (fleupold) [inline .travis.yml]: There is still an open merge conflict

**Review by fleupold — COMMENTED** [review]
Still open merge conflicts + some unrelated changes

File: pepper/Makefile (hunk 5 above)
Comment (josojo) [reply pepper/Makefile]: If, I do not link it, it gives me this error:
test/apps/declarations.h:2:63: fatal error: libsnark/common/default_types/r1cs_ppzksnark_pp.hpp: No such file or directory
 #include <libsnark/common/default_types/r1cs_ppzksnark_pp.hpp>

Comment (fleupold) [reply pepper/Makefile]: Ah right, because you are now depending on libff in the unit tests as well. Makes sense.

**Review by fleupold — APPROVED** [review]
Juhu, if builds pass we are good to merge! 

🏁
